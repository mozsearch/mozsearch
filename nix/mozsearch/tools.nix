{
  rustPlatform,
  runCommandLocal,
  pkgconf,
  cmake,
  openssl,
}: let
  subdir = "tools";
in
  rustPlatform.buildRustPackage {
    pname = "mozsearch-tools";
    version = "unstable";

    src = runCommandLocal "mozsearch-tools-source" {} ''
      mkdir -p $out
      cp -r ${../../tools} $out/tools
      cp -r ${../../deps} $out/deps
    '';

    buildAndTestSubdir = subdir;
    cargoRoot = subdir;

    cargoLock = {
      lockFile = ../../${subdir}/Cargo.lock;
      outputHashes = {
        "traitobject-0.1.0" = "sha256-hnCMcSLblsa0BTRxV/Q7wE1C2KByIf51jYvZGoKFV4M=";
      };
    };

    nativeBuildInputs = [
      pkgconf
      cmake
    ];

    buildInputs = [
      openssl
    ];
  }
