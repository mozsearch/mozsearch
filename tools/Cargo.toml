[package]
name = "tools"
version = "0.1.0"
authors = ["Bill McCloskey <billm@mozilla.com>"]

[dependencies]
chrono = "0.2"
rustc-serialize = "0.3.18"
git2 = "0.7"
lazy_static = "0.2.1"
hyper = "0.10"
log = "0.4.0"
env_logger = "0.5.13"
regex = "1"
ipdl_parser = { path = "./ipdl_parser" }
getopts = "0.2.14"
memmap = "0.5.0"
linkify = "0.2.0"
rls-analysis = "0.17"
rls-data = "0.19"
clap = "2"

# Build release mode with line number info for easier debugging when
# we hit panics in production
[profile.release]
debug = 1
