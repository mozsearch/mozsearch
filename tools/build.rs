fn main() -> Result<(), Box<dyn std::error::Error>> {
    // right now we only use the livegrep proto and not the config proto
    tonic_build::configure().build_server(false).compile(
        &["../deps/livegrep/src/proto/livegrep.proto"],
        &["../deps/livegrep/"],
    )?;
    Ok(())
}
