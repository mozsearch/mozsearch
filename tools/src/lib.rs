#[macro_use]
extern crate lazy_static;
#[macro_use]
extern crate log;
extern crate rustc_serialize;
extern crate git2;
extern crate regex;
extern crate chrono;
extern crate linkify;

pub mod file_format;

pub mod config;
pub mod blame;
pub mod output;
pub mod languages;
pub mod format;
pub mod tokenize;
pub mod links;

pub fn find_source_file(path: &str, files_root: &str, objdir: &str) -> String {
    if path.starts_with("__GENERATED__") {
        return path.replace("__GENERATED__", objdir);
    }
    format!("{}/{}", files_root, path)
}
