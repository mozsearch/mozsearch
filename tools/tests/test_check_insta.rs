use std::{
    io, str,
    sync::{Arc, Mutex, MutexGuard, TryLockError},
};

use serde_json::{from_str, json, Value, to_value};
use tokio::fs::{create_dir_all, read_to_string, write};
use tools::{
    abstract_server::ServerError,
    cmd_pipeline::{build_pipeline, PipelineValues},
    glob_helper::block_in_place_glob_tree,
    templating::builder::build_and_parse_pipeline_explainer,
};
use tracing::dispatcher::Dispatch;
use tracing_subscriber::fmt::MakeWriter;

/// MockWriter and MockMakeWriter is currently taken verbatim from the MIT licensed
/// tracing project that is an existing dependency.  It exists only as a cfg(test)
/// module in
/// https://github.com/tokio-rs/tracing/blob/master/tracing-subscriber/src/fmt/mod.rs
/// hence the need to replicate the code here.
pub struct MockWriter {
    buf: Arc<Mutex<Vec<u8>>>,
}

impl MockWriter {
    pub(crate) fn new(buf: Arc<Mutex<Vec<u8>>>) -> Self {
        Self { buf }
    }

    pub(crate) fn map_error<Guard>(err: TryLockError<Guard>) -> io::Error {
        match err {
            TryLockError::WouldBlock => io::Error::from(io::ErrorKind::WouldBlock),
            TryLockError::Poisoned(_) => io::Error::from(io::ErrorKind::Other),
        }
    }

    pub(crate) fn buf(&self) -> io::Result<MutexGuard<'_, Vec<u8>>> {
        self.buf.try_lock().map_err(Self::map_error)
    }
}

impl io::Write for MockWriter {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        self.buf()?.write(buf)
    }

    fn flush(&mut self) -> io::Result<()> {
        self.buf()?.flush()
    }
}

#[derive(Clone, Default)]
pub(crate) struct MockMakeWriter {
    buf: Arc<Mutex<Vec<u8>>>,
}

impl MockMakeWriter {
    // currently unused
    /*
    pub(crate) fn new(buf: Arc<Mutex<Vec<u8>>>) -> Self {
        Self { buf }
    }

    pub(crate) fn buf(&self) -> MutexGuard<'_, Vec<u8>> {
        self.buf.lock().unwrap()
    }
    */

    pub(crate) fn get_string(&self) -> String {
        let mut buf = self.buf.lock().expect("lock shouldn't be poisoned");
        let string = std::str::from_utf8(&buf[..])
            .expect("formatter should not have produced invalid utf-8")
            .to_owned();
        buf.clear();
        string
    }
}

impl<'a> MakeWriter<'a> for MockMakeWriter {
    type Writer = MockWriter;

    fn make_writer(&'a self) -> Self::Writer {
        MockWriter::new(self.buf.clone())
    }
}

/// Glob-style insta test where we process all of the searchfox-tool command
/// lines under TREE/checks/inputs and output the results of those pipelines to
/// TREE/checks/snapshots using `insta` which provides diff functionality.
///
/// This very dubiously currently relies on having an environment variable
/// CHECK_ROOT defined to tell us where the TREE is.  One might wonder whether
/// this should actually be a test at all or whether it should be its own binary
/// or maybe searchfox-tool should know how to do this or what.
///
/// The reality is that we expect this to be invoked indirectly via
/// `check-index.sh` and never directly triggered via `cargo test`, so... yeah,
/// maybe we could do this in a more clean fashion.  Better opinions accepted!
///
/// `insta` does provide support for binding settings using an `async` function,
/// but its "glob" mechanism does not support `async` so we attempt to reproduce
/// the subset we need from tokio::fs.  (We don't need to use tokio for this,
/// but since we've already started down that road, we stay on the road.  The
/// use of tokio for this is separate from the async limitations on insta's glob
/// which necessitate us doing our own file-finding.)
#[tokio::test(flavor = "multi_thread")]
async fn test_check_glob() -> Result<(), std::io::Error> {
    if let Ok(check_root) = std::env::var("CHECK_ROOT") {
        let explain_template = build_and_parse_pipeline_explainer();

        let input_path = format!("{}/inputs", check_root);
        let snapshot_root_path = format!("{}/snapshots", check_root);

        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);

        // ## Figure out the list of input files
        let input_names = block_in_place_glob_tree(&input_path, "**/*");

        for (rel_path, filename) in input_names {
            let input_path = format!("{}/inputs/{}{}", check_root, rel_path, filename);
            settings.set_input_file(&input_path);
            let snapshot_path = format!("{}/{}", snapshot_root_path, rel_path);
            create_dir_all(snapshot_path.clone()).await?;
            settings.set_snapshot_path(snapshot_path);
            settings.set_snapshot_suffix(filename.clone());

            let make_writer = MockMakeWriter::default();
            let subscriber = tracing_subscriber::fmt()
                .json()
                .with_env_filter("tools=trace")
                .with_current_span(false)
                .with_writer(make_writer.clone())
                .finish();

            let dispatcher = Dispatch::new(subscriber);
            let _trace_guard = tracing::dispatcher::set_default(&dispatcher);
            let mut server_kind = "unknown".to_string();

            settings
                .bind_async(async {
                    let command = read_to_string(input_path).await.unwrap();

                    let pipeline = match build_pipeline(&"searchfox-tool", &command) {
                        Ok((pipeline, _)) => pipeline,
                        Err(err) => {
                            insta::assert_snapshot!(format!("Pipeline Build Error: {:?}", err));
                            return;
                        }
                    };
                    server_kind = pipeline.server_kind.clone();
                    let results = pipeline.run(true).await;

                    // TODO: In theory we should perhaps block_in_place here, but also it doesn't
                    // matter.
                    match results {
                        Ok(PipelineValues::Void) => {
                            insta::assert_snapshot!("void");
                        }
                        Ok(PipelineValues::IdentifierList(il)) => {
                            insta::assert_json_snapshot!(json!(il.identifiers));
                        }
                        Ok(PipelineValues::SymbolList(sl)) => {
                            insta::assert_json_snapshot!(&to_value(sl).unwrap());
                        }
                        Ok(PipelineValues::SymbolCrossrefInfoList(scil)) => {
                            // We used to previously just turn this into a list of
                            // just the crossref values, but we now have important
                            // metadata.
                            insta::assert_json_snapshot!(&to_value(scil).unwrap());
                        }
                        Ok(PipelineValues::SymbolGraphCollection(sgc)) => {
                            insta::assert_json_snapshot!(sgc.to_json());
                        }
                        Ok(PipelineValues::FlattenedResultsBundle(frb)) => {
                            insta::assert_json_snapshot!(&to_value(frb).unwrap());
                        }
                        Ok(PipelineValues::HtmlExcerpts(he)) => {
                            let mut aggr_str = String::new();
                            for file_excerpts in he.by_file {
                                for str in file_excerpts.excerpts {
                                    aggr_str += &str;
                                }
                            }
                            insta::assert_snapshot!(&aggr_str);
                        }
                        Ok(PipelineValues::TextFile(fb)) => {
                            insta::assert_snapshot!(&fb.contents);
                        }
                        Ok(PipelineValues::JsonRecords(jr)) => {
                            let mut json_results = vec![];
                            for file_records in jr.by_file {
                                json_results.extend(file_records.records);
                            }

                            insta::assert_json_snapshot!(&json_results);
                        }
                        Ok(PipelineValues::JsonValue(jv)) => {
                            insta::assert_json_snapshot!(&jv.value);
                        }
                        Ok(PipelineValues::FileMatches(fm)) => {
                            insta::assert_json_snapshot!(&to_value(fm).unwrap());
                        }
                        Ok(PipelineValues::TextMatches(tm)) => {
                            insta::assert_json_snapshot!(&to_value(tm).unwrap());
                        }
                        Err(ServerError::Unsupported) => {
                            // We're intentionally skipping doing anything here.
                            // Our assumption is that this error will only be
                            // returned in cases like the local index server
                            // being unable to handle "query" commands.
                        }
                        Err(err) => {
                            insta::assert_snapshot!(format!("Pipeline Error: {:?}", err));
                        }
                    }
                })
                .await;

            let log_values: Vec<Value> = make_writer
                .get_string()
                .lines()
                .map(|s| from_str(s).unwrap())
                .collect();

            let explain_dir = format!("{}/explanations/{}", check_root, rel_path);
            create_dir_all(explain_dir.clone()).await?;
            let explain_path = format!("{}{}-{}.md", explain_dir, filename, server_kind);

            let globals = liquid::object!({
                "logs": log_values,
            });

            let output = explain_template.render(&globals).unwrap();
            write(explain_path, output).await?;
        }
    }

    Ok(())
}
