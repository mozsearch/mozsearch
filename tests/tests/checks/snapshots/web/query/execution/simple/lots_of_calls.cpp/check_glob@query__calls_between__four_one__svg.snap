---
source: tests/test_check_insta.rs
expression: "&to_value(grb).unwrap()"
---
{
  "graphs": [
    {
      "graph": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 2.43.0 (0)\n -->\n<!-- Title: g Pages: 1 -->\n<svg width=\"510pt\" height=\"260pt\"\n viewBox=\"0.00 0.00 510.00 260.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 256)\" data-symbols=\"g\">\n<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-256 506,-256 506,4 -4,4\"/>\n<!-- _ZN9CallerOne18one_calls_two_leftEP9CallerTwoP11CallerThreeP10CallerFour -->\n<g id=\"node1\" class=\"node\" data-symbols=\"_ZN9CallerOne18one_calls_two_leftEP9CallerTwoP11CallerThreeP10CallerFour\">\n<polygon fill=\"none\" stroke=\"black\" points=\"215,-252 24,-252 24,-216 215,-216 215,-252\"/>\n<text text-anchor=\"middle\" x=\"119.5\" y=\"-231.5\" font-family=\"Courier New\" font-size=\"10.00\">CallerOne::one_calls_two_left</text>\n</g>\n<!-- _ZN9CallerTwo26two_left_calls_three_nexusEP11CallerThreeP10CallerFour -->\n<g id=\"node2\" class=\"node\" data-symbols=\"_ZN9CallerTwo26two_left_calls_three_nexusEP11CallerThreeP10CallerFour\">\n<polygon fill=\"none\" stroke=\"black\" points=\"239,-180 0,-180 0,-144 239,-144 239,-180\"/>\n<text text-anchor=\"middle\" x=\"119.5\" y=\"-159.5\" font-family=\"Courier New\" font-size=\"10.00\">CallerTwo::two_left_calls_three_nexus</text>\n</g>\n<!-- _ZN9CallerOne18one_calls_two_leftEP9CallerTwoP11CallerThreeP10CallerFour&#45;&gt;_ZN9CallerTwo26two_left_calls_three_nexusEP11CallerThreeP10CallerFour -->\n<g id=\"edge1\" class=\"edge\" data-symbols=\"_ZN9CallerOne18one_calls_two_leftEP9CallerTwoP11CallerThreeP10CallerFour&#45;&gt;_ZN9CallerTwo26two_left_calls_three_nexusEP11CallerThreeP10CallerFour\">\n<path fill=\"none\" stroke=\"black\" d=\"M119.5,-215.7C119.5,-207.98 119.5,-198.71 119.5,-190.11\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"123,-190.1 119.5,-180.1 116,-190.1 123,-190.1\"/>\n</g>\n<!-- _ZN11CallerThree11three_nexusEP10CallerFour -->\n<g id=\"node3\" class=\"node\" data-symbols=\"_ZN11CallerThree11three_nexusEP10CallerFour\">\n<polygon fill=\"none\" stroke=\"black\" points=\"330,-108 169,-108 169,-72 330,-72 330,-108\"/>\n<text text-anchor=\"middle\" x=\"249.5\" y=\"-87.5\" font-family=\"Courier New\" font-size=\"10.00\">CallerThree::three_nexus</text>\n</g>\n<!-- _ZN9CallerTwo26two_left_calls_three_nexusEP11CallerThreeP10CallerFour&#45;&gt;_ZN11CallerThree11three_nexusEP10CallerFour -->\n<g id=\"edge2\" class=\"edge\" data-symbols=\"_ZN9CallerTwo26two_left_calls_three_nexusEP11CallerThreeP10CallerFour&#45;&gt;_ZN11CallerThree11three_nexusEP10CallerFour\">\n<path fill=\"none\" stroke=\"black\" d=\"M151.3,-143.88C168.61,-134.55 190.22,-122.92 208.76,-112.94\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"210.43,-116.01 217.58,-108.19 207.11,-109.85 210.43,-116.01\"/>\n</g>\n<!-- _ZN10CallerFour9four_leftEv -->\n<g id=\"node4\" class=\"node\" data-symbols=\"_ZN10CallerFour9four_leftEv\">\n<polygon fill=\"none\" stroke=\"black\" points=\"239,-36 96,-36 96,0 239,0 239,-36\"/>\n<text text-anchor=\"middle\" x=\"167.5\" y=\"-15.5\" font-family=\"Courier New\" font-size=\"10.00\">CallerFour::four_left</text>\n</g>\n<!-- _ZN11CallerThree11three_nexusEP10CallerFour&#45;&gt;_ZN10CallerFour9four_leftEv -->\n<g id=\"edge3\" class=\"edge\" data-symbols=\"_ZN11CallerThree11three_nexusEP10CallerFour&#45;&gt;_ZN10CallerFour9four_leftEv\">\n<path fill=\"none\" stroke=\"black\" d=\"M229.23,-71.7C219,-62.97 206.44,-52.24 195.32,-42.75\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"197.41,-39.94 187.54,-36.1 192.87,-45.26 197.41,-39.94\"/>\n</g>\n<!-- _ZN10CallerFour10four_rightEv -->\n<g id=\"node7\" class=\"node\" data-symbols=\"_ZN10CallerFour10four_rightEv\">\n<polygon fill=\"none\" stroke=\"black\" points=\"406,-36 257,-36 257,0 406,0 406,-36\"/>\n<text text-anchor=\"middle\" x=\"331.5\" y=\"-15.5\" font-family=\"Courier New\" font-size=\"10.00\">CallerFour::four_right</text>\n</g>\n<!-- _ZN11CallerThree11three_nexusEP10CallerFour&#45;&gt;_ZN10CallerFour10four_rightEv -->\n<g id=\"edge6\" class=\"edge\" data-symbols=\"_ZN11CallerThree11three_nexusEP10CallerFour&#45;&gt;_ZN10CallerFour10four_rightEv\">\n<path fill=\"none\" stroke=\"black\" d=\"M269.77,-71.7C280,-62.97 292.56,-52.24 303.68,-42.75\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"306.13,-45.26 311.46,-36.1 301.59,-39.94 306.13,-45.26\"/>\n</g>\n<!-- _ZN9CallerOne19one_calls_two_rightEP9CallerTwoP11CallerThreeP10CallerFour -->\n<g id=\"node5\" class=\"node\" data-symbols=\"_ZN9CallerOne19one_calls_two_rightEP9CallerTwoP11CallerThreeP10CallerFour\">\n<polygon fill=\"none\" stroke=\"black\" points=\"478,-252 281,-252 281,-216 478,-216 478,-252\"/>\n<text text-anchor=\"middle\" x=\"379.5\" y=\"-231.5\" font-family=\"Courier New\" font-size=\"10.00\">CallerOne::one_calls_two_right</text>\n</g>\n<!-- _ZN9CallerTwo27two_right_calls_three_nexusEP11CallerThreeP10CallerFour -->\n<g id=\"node6\" class=\"node\" data-symbols=\"_ZN9CallerTwo27two_right_calls_three_nexusEP11CallerThreeP10CallerFour\">\n<polygon fill=\"none\" stroke=\"black\" points=\"502,-180 257,-180 257,-144 502,-144 502,-180\"/>\n<text text-anchor=\"middle\" x=\"379.5\" y=\"-159.5\" font-family=\"Courier New\" font-size=\"10.00\">CallerTwo::two_right_calls_three_nexus</text>\n</g>\n<!-- _ZN9CallerOne19one_calls_two_rightEP9CallerTwoP11CallerThreeP10CallerFour&#45;&gt;_ZN9CallerTwo27two_right_calls_three_nexusEP11CallerThreeP10CallerFour -->\n<g id=\"edge4\" class=\"edge\" data-symbols=\"_ZN9CallerOne19one_calls_two_rightEP9CallerTwoP11CallerThreeP10CallerFour&#45;&gt;_ZN9CallerTwo27two_right_calls_three_nexusEP11CallerThreeP10CallerFour\">\n<path fill=\"none\" stroke=\"black\" d=\"M379.5,-215.7C379.5,-207.98 379.5,-198.71 379.5,-190.11\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"383,-190.1 379.5,-180.1 376,-190.1 383,-190.1\"/>\n</g>\n<!-- _ZN9CallerTwo27two_right_calls_three_nexusEP11CallerThreeP10CallerFour&#45;&gt;_ZN11CallerThree11three_nexusEP10CallerFour -->\n<g id=\"edge5\" class=\"edge\" data-symbols=\"_ZN9CallerTwo27two_right_calls_three_nexusEP11CallerThreeP10CallerFour&#45;&gt;_ZN11CallerThree11three_nexusEP10CallerFour\">\n<path fill=\"none\" stroke=\"black\" d=\"M347.7,-143.88C330.39,-134.55 308.78,-122.92 290.24,-112.94\"/>\n<polygon fill=\"black\" stroke=\"black\" points=\"291.89,-109.85 281.42,-108.19 288.57,-116.01 291.89,-109.85\"/>\n</g>\n</g>\n</svg>\n"
    }
  ],
  "overloads_hit": [],
  "symbols": {
    "_ZN10CallerFour10four_rightEv": {
      "structured": 1,
      "pretty": "CallerFour::four_right",
      "sym": "_ZN10CallerFour10four_rightEv",
      "kind": "method",
      "parentsym": "T_CallerFour",
      "implKind": "",
      "sizeBytes": null,
      "bindingSlots": [],
      "supers": [],
      "methods": [],
      "fields": [],
      "overrides": [],
      "props": [
        "instance",
        "user"
      ]
    },
    "_ZN10CallerFour9four_leftEv": {
      "structured": 1,
      "pretty": "CallerFour::four_left",
      "sym": "_ZN10CallerFour9four_leftEv",
      "kind": "method",
      "parentsym": "T_CallerFour",
      "implKind": "",
      "sizeBytes": null,
      "bindingSlots": [],
      "supers": [],
      "methods": [],
      "fields": [],
      "overrides": [],
      "props": [
        "instance",
        "user"
      ]
    },
    "_ZN11CallerThree11three_nexusEP10CallerFour": {
      "structured": 1,
      "pretty": "CallerThree::three_nexus",
      "sym": "_ZN11CallerThree11three_nexusEP10CallerFour",
      "kind": "method",
      "parentsym": "T_CallerThree",
      "implKind": "",
      "sizeBytes": null,
      "bindingSlots": [],
      "supers": [],
      "methods": [],
      "fields": [],
      "overrides": [],
      "props": [
        "instance",
        "user"
      ]
    },
    "_ZN9CallerOne18one_calls_two_leftEP9CallerTwoP11CallerThreeP10CallerFour": {
      "structured": 1,
      "pretty": "CallerOne::one_calls_two_left",
      "sym": "_ZN9CallerOne18one_calls_two_leftEP9CallerTwoP11CallerThreeP10CallerFour",
      "kind": "method",
      "parentsym": "T_CallerOne",
      "implKind": "",
      "sizeBytes": null,
      "bindingSlots": [],
      "supers": [],
      "methods": [],
      "fields": [],
      "overrides": [],
      "props": [
        "instance",
        "user"
      ]
    },
    "_ZN9CallerOne19one_calls_two_rightEP9CallerTwoP11CallerThreeP10CallerFour": {
      "structured": 1,
      "pretty": "CallerOne::one_calls_two_right",
      "sym": "_ZN9CallerOne19one_calls_two_rightEP9CallerTwoP11CallerThreeP10CallerFour",
      "kind": "method",
      "parentsym": "T_CallerOne",
      "implKind": "",
      "sizeBytes": null,
      "bindingSlots": [],
      "supers": [],
      "methods": [],
      "fields": [],
      "overrides": [],
      "props": [
        "instance",
        "user"
      ]
    },
    "_ZN9CallerTwo26two_left_calls_three_nexusEP11CallerThreeP10CallerFour": {
      "structured": 1,
      "pretty": "CallerTwo::two_left_calls_three_nexus",
      "sym": "_ZN9CallerTwo26two_left_calls_three_nexusEP11CallerThreeP10CallerFour",
      "kind": "method",
      "parentsym": "T_CallerTwo",
      "implKind": "",
      "sizeBytes": null,
      "bindingSlots": [],
      "supers": [],
      "methods": [],
      "fields": [],
      "overrides": [],
      "props": [
        "instance",
        "user"
      ]
    },
    "_ZN9CallerTwo27two_right_calls_three_nexusEP11CallerThreeP10CallerFour": {
      "structured": 1,
      "pretty": "CallerTwo::two_right_calls_three_nexus",
      "sym": "_ZN9CallerTwo27two_right_calls_three_nexusEP11CallerThreeP10CallerFour",
      "kind": "method",
      "parentsym": "T_CallerTwo",
      "implKind": "",
      "sizeBytes": null,
      "bindingSlots": [],
      "supers": [],
      "methods": [],
      "fields": [],
      "overrides": [],
      "props": [
        "instance",
        "user"
      ]
    }
  }
}
